evolver/device.py

```diff
@@ -71,19 +72,27 @@ def schema(self):
             "controllers": [{"kind": str(type(a)), "config": a.Config.model_json_schema()} for a in self.controllers],
         }
 
+    def _loop_exception_wrapper(self, callable, message="uknown"):
+        try:
+            callable()
+        except Exception as e:
+            if self.raise_loop_exceptions:
+                raise e
+            self.logger.error(f"Error in loop: {message}: {e}")
+
     def read_state(self):
         for name, device in self.sensors.items():
-            device.read()
+            self._loop_exception_wrapper(device.read, f"reading device {name}")
```

Comment:
Wouldn't this imply that the caller would be ok with possibly receiving garbage? This could trigger a down stream exception that would only raise anyhow.

Reply:
maybe, but if they do receive something, that something would have been added to output by the driver and it would be responsible for either delivering garbage or not, I think. And yeah, if the consumer of it raised that would be handled and logged in the controller section.